src/app-layer-template.c

```diff
@@ -0,0 +1,419 @@
+    TemplateTransaction *templatetx = tx;
+    if (templatetx->request_buffer != NULL) {
+        SCFree(templatetx->request_buffer);
+    }
+    if (templatetx->response_buffer != NULL) {
+        SCFree(templatetx->response_buffer);
+    }
+    SCFree(tx);
+}
+
+static void *TemplateStateAlloc(void)
+{
+    SCLogNotice("Allocating template state.");
+    TemplateState *state = SCCalloc(1, sizeof(TemplateState));
+    if (unlikely(state == NULL)) {
+        return NULL;
+    }
+    TAILQ_INIT(&state->tx_list);
+    return state;
+}
+
+static void TemplateStateFree(void *state)
+{
+    TemplateState *template_state = state;
+    TemplateTransaction *tx;
+    SCLogNotice("Freeing template state.");
+    while ((tx = TAILQ_FIRST(&template_state->tx_list)) != NULL) {
+        TAILQ_REMOVE(&template_state->tx_list, tx, next);
+        TemplateTxFree(tx);
+    }
+    SCFree(template_state);
+}
+
+/**
+ * \brief Callback from the application layer to have a transaction freed.
+ *
+ * \param state a void pointer to the TemplateState object.
+ * \param tx_id the transaction ID to free.
+ */
+static void TemplateStateTxFree(void *state, uint64_t tx_id)
+{
+    TemplateState *echo = state;
+    TemplateTransaction *tx = NULL, *ttx;
+
+    SCLogNotice("Freeing transaction %"PRIu64, tx_id);
+
+    TAILQ_FOREACH_SAFE(tx, &echo->tx_list, next, ttx) {
+
+        /* Continue if this is not the transaction we are looking
+         * for. */
+        if (tx->tx_id != tx_id) {
+            continue;
+        }
+
+        /* Remove and free the transaction. */
+        TAILQ_REMOVE(&echo->tx_list, tx, next);
+        TemplateTxFree(tx);
+        break;
+    }
+}
+
+/**
+ * \brief Probe the input to see if it looks like echo.
+ *
+ * \retval ALPROTO_TEMPLATE if it looks like echo, otherwise
+ *     ALPROTO_UNKNOWN.
+ */
+static AppProto TemplateProbingParser(uint8_t *input, uint32_t input_len,
+    uint32_t *offset)
+{
+    /* Very simple test - if there is input, this is echo. */
+    if (input_len >= TEMPLATE_MIN_FRAME_LEN) {
+        SCLogNotice("Detected as ALPROTO_TEMPLATE.");
+        return ALPROTO_TEMPLATE;
+    }
+
+    SCLogNotice("Protocol not detected as ALPROTO_TEMPLATE.");
+    return ALPROTO_UNKNOWN;
+}
+
+static int TemplateParseRequest(Flow *f, void *state, AppLayerParserState *pstate,
+    uint8_t *input, uint32_t input_len, void *local_data)
+{
+    TemplateState *echo = state;
+
+    SCLogNotice("Parsing echo request.");
+
+    /* Normally you would parse out data here and store it in the
+     * transcation object, but as this is echo, we'll just record the
+     * request data. */
+
+    /* Allocate a transaction. */
+    TemplateTransaction *tx = TemplateTxAlloc(echo);
+    if (unlikely(tx == NULL)) {
+        SCLogNotice("Failed to allocate new Template tx.");
+        goto end;
+    }
+    SCLogNotice("Allocated Template tx %"PRIu64".", tx->tx_id);
+    
+    /* Make a copy of the request. */
+    tx->request_buffer = SCCalloc(1, input_len);
```

Comment:
Its a little out of scope here, as we just allocated the transaction, instead of looked one up.  Will comment on the matter though.